docs/source/custom-collections.rst

```diff
@@ -0,0 +1,326 @@
+Custom Collections
+==================
+
+For many problems the built-in dask collections (``dask.array``,
+``dask.dataframe``, ``dask.bag``, and ``dask.delayed``) are sufficient. For
+cases where they aren't it's possible to create your own dask collections. Here
+we describe the required methods to fullfill the dask collection interface.
+
+.. note:: This is considered an advanced feature. For most cases the built-in
+          collections are probably sufficient.
+
+Before reading this you should read and underestand:
+
+- :doc:`overview <graphs>`
+- :doc:`graph specification <spec>`
+- :doc:`custom graphs <custom-graphs>`
+
+The Dask Collection Interface
+-----------------------------
+
+To create your own dask collection, you need to fullfill the following
+interface. Note that there is no required base class:
+
+.. method:: __dask_graph__(self)
+
+    The dask graph.
+
+    Returns
+    -------
+    dsk : MutableMapping, None
+        The dask graph. If ``None``, this instance will not be interpreted as a
+        dask collection, and none of the remaining interface methods will be
+        called.
+
+
+.. method:: __dask_keys__(self)
+
+    The output keys for the dask graph.
+
+    Returns
+    -------
+    keys : list
+        A possibly nested list of keys that represent the outputs of the graph.
+        After computation, the results will be returned in the same layout,
+        with the keys replaced with their corresponding outputs.
+
+
+.. staticmethod:: __dask_optimize__(dsk, keys, \*\*kwargs)
+
+    Given a graph and keys, return a new opimized graph.
+
+    This method can be either a ``staticmethod`` or a ``classmethod``, but not
+    an instancemethod.
+
+    If not implemented, defaults to returning the graph unchanged.
+
+    Parameters
+    ----------
+    dsk : MutableMapping
+    keys : list
+    \*\*kwargs
+        Extra keyword arguments forwarded from the call to ``compute`` or
+        ``persist``. Can be used or ignored as needed.
```

Comment:
For xarray, I suppose we should simply forward to the optimize method from dask.array? e.g., `da.Array.__dask_optimize__(dsk, **kwargs)`?

Is there a guarantee that `keys` refers to only a single dask object of this type, or could they be taken from a collection of objects of this type or even objects of different dask types? (I think the last one, but that's not immediately obvious.)

Some pseudo-code for how this is called could by the scheduler could be helpful in thinking this all through.

Reply:
For xarray, I suppose we should simply forward to the optimize method from dask.array? e.g., `da.Array.__dask_optimize__(dsk, **kwargs)`. Could also do something like:

```python
class YourClass(...):
    @property
    def __dask_optimize__(self):
        return da.Array.__dask_optimize__
```

> Is there a guarantee that keys refers to only a single dask object of this type, or could they be taken from a collection of objects of this type or even objects of different dask types? (I think the last one, but that's not immediately obvious.)

The latter. I added a section documenting how these methods are used in the dask core methods (e.g. `compute`) that hopefully clarifies this. Also clarified the method docs.